.github/workflows/main.yml

```diff
@@ -21,29 +21,29 @@ jobs:
         run: pip install -r requirements.txt && python download_server.py
 
       - name: Setup Java JDK
-        uses: actions/setup-java@v3
+        uses: actions/setup-java@v4
         with:
           distribution: 'temurin'
           java-version: 17
           check-latest: true
 
       - name: Validate Gradle Wrapper
-        uses: gradle/wrapper-validation-action@v1
+        uses: gradle/wrapper-validation-action@v2.1.1
 
       # Wooyeah
       - name: Read last snapshot
         id: last_snapshot
-        uses: juliangruber/read-file-action@v1
+        uses: juliangruber/read-file-action@v1.1.7
```

Comment:
Revert this part as my request above applies here along with the same parts on the other two.